'npm:graphql-tag' does not provide an export named 'gql'

This is my TS script:

```
import { gql } from "npm:graphql-tag";

gql;
```

When I run it with `deno run -A main.ts`, I get this error:

```
error: Uncaught SyntaxError: The requested module 'npm:graphql-tag' does not provide an export named 'gql'
import { gql } from "npm:graphql-tag";
```

This is anormal because the graphql-tag does export a gql object.

Experienced the same issue when making a similar effort to that of OP.

I discovered that I was able to (at least partially) use @dagger.io/dagger by importing it from esm.sh as in the following:

```typescript
import Client, { connect } from "https://esm.sh/@dagger.io/dagger@0.6.0";

import process from 'node:process';

connect(async (client: Client) => {

  const node = await client.container().from("node:16").exec(["node", "-v"]);

  const version = await node.stdout().contents();

  console.log("Hello from Dagger and Node " + version.contents);

}, { LogOutput: process.stdout });
```

You will find that `deno run -A main.ts` now fails with a different error:

<details><summary><code>Uncaught InitEngineSessionBinaryError: failed to download dagger cli binary: TypeError: d.pipe is not a function</code></summary>

```
error: Uncaught InitEngineSessionBinaryError: failed to download dagger cli binary: TypeError: d.pipe is not a function
    at G.<anonymous> (https://esm.sh/v124/@dagger.io/dagger@0.6.0/deno/dagger.mjs:8:23837)
    at Generator.throw (<anonymous>)
    at T (https://esm.sh/v124/@dagger.io/dagger@0.6.0/deno/dagger.mjs:8:22256)
    at eventLoopTick (ext:core/01_core.js:181:11)
Caused by: TypeError: d.pipe is not a function
    at https://esm.sh/v124/@dagger.io/dagger@0.6.0/deno/dagger.mjs:9:679
    at new Promise (<anonymous>)
    at G.<anonymous> (https://esm.sh/v124/@dagger.io/dagger@0.6.0/deno/dagger.mjs:9:618)
    at Generator.next (<anonymous>)
    at l (https://esm.sh/v124/@dagger.io/dagger@0.6.0/deno/dagger.mjs:8:22205)
    at eventLoopTick (ext:core/01_core.js:181:11)
```

</details>

... however, this can be worked around using the [Dagger CLI](https://docs.dagger.io/cli) with `dagger run deno run -A main.ts`. This is because `dagger run` sets up a session which it provides to the Dagger SDK using environment variables, whereas otherwise the SDK will attempt to download and run the Dagger CLI itself to create the session. The listed error appears to be related to this mechanism, which is not needed when invoking Deno from within the Dagger CLI.

P.S. Dagger now has an issue for Deno support: https://github.com/dagger/dagger/issues/4368

--- a/libs/node_resolver/analyze.rs
+++ b/libs/node_resolver/analyze.rs
@@ -221,7 +221,7 @@ impl<
 
     if !analysis.member_reexports.is_empty() {
       let mut errors = Vec::new();
-      self
+      let fallback_reexports = self
         .resolve_member_reexports(
           entry_specifier,
           &analysis.member_reexports,
@@ -229,6 +229,18 @@ impl<
           &mut errors,
         )
         .await;
+      // Members whose attached names couldn't be determined statically
+      // fall back to a wholesale re-export of the inner module.
+      if !fallback_reexports.is_empty() {
+        self
+          .analyze_reexports(
+            entry_specifier,
+            fallback_reexports,
+            &mut all_exports,
+            &mut errors,
+          )
+          .await;
+      }
       if !errors.is_empty() {
         errors.sort_by_cached_key(|e| e.to_string());
         return Err(TranslateCjsToEsmError::ExportAnalysis(errors.remove(0)));
@@ -245,6 +257,13 @@ impl<
   /// strictly narrower than treating `X` as a wildcard re-export: only
   /// names the inner module statically attaches to the specific member
   /// are advertised, so unrelated names from `X` don't leak through.
+  ///
+  /// When the attached names can't be determined statically (e.g.
+  /// graphql-tag@2's UMD wraps its `exports.gql = …` / `gql.* = …`
+  /// assignments inside the factory IIFE and builds the value through a
+  /// namespace alias), the inner specifier is returned so the caller can
+  /// fall back to a wholesale re-export, matching Node's behavior for
+  /// `module.exports = require(X).Y`.
   #[allow(
     clippy::needless_lifetimes,
     reason = "explicit lifetimes improve clarity"
@@ -255,7 +274,8 @@ impl<
     member_reexports: &[CjsMemberReExport],
     all_exports: &mut BTreeSet<String>,
     errors: &mut Vec<JsErrorBox>,
-  ) {
+  ) -> Vec<String> {
+    let mut fallback_reexports = Vec::new();
     for entry in member_reexports {
       let result = self
         .resolve(
@@ -290,7 +310,12 @@ impl<
         .await
       {
         Ok(Some(props)) => props,
-        Ok(None) => continue,
+        // Couldn't statically narrow to the member's attached names;
+        // fall back to re-exporting the inner module wholesale.
+        Ok(None) => {
+          fallback_reexports.push(entry.specifier.clone());
+          continue;
+        }
         Err(err) => {
           errors.push(err);
           continue;
@@ -303,6 +328,7 @@ impl<
         all_exports.insert(prop);
       }
     }
+    fallback_reexports
   }
 
   #[allow(
@@ -430,7 +456,7 @@ impl<
           }
 
           if !analysis.member_reexports.is_empty() {
-            self
+            let fallback_reexports = self
               .resolve_member_reexports(
                 &reexport_specifier,
                 &analysis.member_reexports,
@@ -438,6 +464,16 @@ impl<
                 errors,
               )
               .await;
+            // Members that couldn't be narrowed statically fall back to a
+            // wholesale re-export, fed back through the same loop.
+            if !fallback_reexports.is_empty() {
+              handle_reexports(
+                reexport_specifier.clone(),
+                fallback_reexports,
+                &mut analyze_futures,
+                errors,
+              );
+            }
           }
 
           all_exports.extend(

--- a/tests/registry/npm/@denotest/cjs-module-exports-require-member-umd/1.0.0/lib/inner.umd.js
+++ b/tests/registry/npm/@denotest/cjs-module-exports-require-member-umd/1.0.0/lib/inner.umd.js
@@ -1,0 +1,26 @@
+(function (global, factory) {
+  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
+  typeof define === 'function' && define.amd ? define(['exports'], factory) :
+  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.gqlmod = {}));
+}(this, (function (exports) {
+  'use strict';
+  function gql() { return "gql-result"; }
+  function resetCaches() { return "reset"; }
+  function disableFragmentWarnings() { return "disabled"; }
+  var extras = {
+    gql: gql,
+    resetCaches: resetCaches,
+    disableFragmentWarnings: disableFragmentWarnings,
+  };
+  (function (gql_1) {
+    gql_1.gql = extras.gql;
+    gql_1.resetCaches = extras.resetCaches;
+    gql_1.disableFragmentWarnings = extras.disableFragmentWarnings;
+  })(gql || (gql = {}));
+  gql["default"] = gql;
+  exports.default = gql;
+  exports.gql = gql;
+  exports.resetCaches = resetCaches;
+  exports.disableFragmentWarnings = disableFragmentWarnings;
+  Object.defineProperty(exports, '__esModule', { value: true });
+})));

--- a/tests/registry/npm/@denotest/cjs-module-exports-require-member-umd/1.0.0/main.js
+++ b/tests/registry/npm/@denotest/cjs-module-exports-require-member-umd/1.0.0/main.js
@@ -1,0 +1,3 @@
+// Mirrors graphql-tag@2's main entry exactly: re-export a single named
+// property of an inner UMD module as the whole `module.exports`.
+module.exports = require('./lib/inner.umd.js').gql;

--- a/tests/registry/npm/@denotest/cjs-module-exports-require-member-umd/1.0.0/package.json
+++ b/tests/registry/npm/@denotest/cjs-module-exports-require-member-umd/1.0.0/package.json
@@ -1,0 +1,5 @@
+{
+  "name": "@denotest/cjs-module-exports-require-member-umd",
+  "version": "1.0.0",
+  "main": "main.js"
+}

--- a/tests/specs/npm/cjs_module_exports_require_member_umd/__test__.jsonc
+++ b/tests/specs/npm/cjs_module_exports_require_member_umd/__test__.jsonc
@@ -1,0 +1,4 @@
+{
+  "args": "run -A --quiet main.ts",
+  "output": "main.out"
+}

--- a/tests/specs/npm/cjs_module_exports_require_member_umd/main.out
+++ b/tests/specs/npm/cjs_module_exports_require_member_umd/main.out
@@ -1,0 +1,3 @@
+gql-result
+reset
+disabled

--- a/tests/specs/npm/cjs_module_exports_require_member_umd/main.ts
+++ b/tests/specs/npm/cjs_module_exports_require_member_umd/main.ts
@@ -1,0 +1,16 @@
+// Regression test for #16708: graphql-tag@2's main entry is
+//   module.exports = require("./lib/graphql-tag.umd.js").gql;
+// where the inner module is a UMD bundle that attaches its named
+// exports inside the factory IIFE through a namespace alias. The
+// member's attached names can't be narrowed statically, so the
+// analyzer falls back to re-exporting the inner module wholesale and
+// `import { gql }` resolves (matching Node).
+import {
+  disableFragmentWarnings,
+  gql,
+  resetCaches,
+} from "npm:@denotest/cjs-module-exports-require-member-umd";
+
+console.log(gql());
+console.log(resetCaches());
+console.log(disableFragmentWarnings());

